tools/set_zones/zbc_set_zones.8
.\"  SPDX-License-Identifier: LGPL-3.0-or-later
.\"  SPDX-FileCopyrightText: 2020, Western Digital Corporation or its affiliates.
.\"  Written by Damien Le Moal <damien.lemoal@wdc.com>
.\"
.TH ZBC 8
.SH NAME
zbc_set_zones \- Initialize a file or device for ZBC emulation

.SH SYNOPSIS
.B zbc_set_zones
[options]
.IR device
.IR command
.IR command_arguments
.B zbc_set_zones
[options]
.IR file
.IR command
.IR command_arguments

.SH DESCRIPTION
.B zbc_set_zones
is used to initialize a regular file or a regular block device to use it
as an emulated ZBC device.

.PP
The
.I device
argument must be the pathname to a regular block device file (e.g.,
.IR /dev/sdb "),"
or to a regular file that is not empty. The size of the file is used
to determine the capacity of the emulated device.

.PP
The
.I command
and
.I command_arguments
can have one of the following form.

.TP
.BR set_sz " " \fIcnv_zone_size\fP " " \fIzone_size\fP
Initialize the emulated device by specifying the total capacity to use for
conventional zones and the size of zones. The argument \fIcnv_zone_size\fP
specifies the total capacity of conventional zones in MiB. The argument
\fIzone_size\fP specifies the size of all zones in MiB.
.TP
.BR set_ps " " \fIcnv_zone_size_ratio\fP " " \fIzone_size\fP
Initialize the emulated device by specifying the capacity ratio used for
conventional zones and the size of zones. The argument \fIcnv_zone_size_ratio\fP
specifies the percentage of the total capacity to use for conventional zones.
The argument \fIzone_size\fP specifies the size of all zones in MiB.

.RE
If a regular file is used, the total number of zones is automatically
determined from the size of the file. In the case of a regular block device,
the device capacity is used to calculate the total number of zones that
will be emulated.

.SH OPTIONS
The following options can be specified.
.TP
.BR \-h , " \-\-help"
Display a usage help message and exit.
.TP
.BR \-v
Verbose mode (for debugging problems).

.SH AUTHOR
.nf
Damien Le Moal <damien.lemoal@wdc.com>
.fi

.SH SEE ALSO
.na
.BR zbc_set_write_ptr (8)
.ad

.SH AVAILABILITY
The \fBzbc_set_zones\fP utility is part of the \fBlibzbc\fP library available
from
.br
.UR https://\:github.com\:/westerndigitalcorporation\:/libzbc
.UE .


tools/set_zones/zbc_set_zones.c
// SPDX-License-Identifier: LGPL-3.0-or-later
/*
 * This file is part of libzbc.
 *
 * Copyright (C) 2009-2014, HGST, Inc. All rights reserved.
 * Copyright (C) 2016, Western Digital. All rights reserved.
 * Copyright (C) 2020 Western Digital COrporation or its affiliates.
 *
 * Author: Damien Le Moal (damien.lemoal@wdc.com)
 *         Christophe Louargant (christophe.louargant@wdc.com)
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <libgen.h>

#include <libzbc/zbc.h>

#include <zbc_private.h>

static int zbc_set_zones_usage(char *prog)
{
	printf("Usage: %s [options] <dev> <command> <command arguments>\n"
	       "Options:\n"
	       "  -h | --help : Display this help message and exit\n"
	       "  -v          : Verbose mode\n"
	       "Commands:\n"
	       "  set_sz <conv zone size (MB)> <zone size (MiB)> :\n"
	       "      Specify the total size in MiB of all conventional\n"
	       "      zones and the size in MiB of zones\n"
	       "  set_ps <conv zone size (%%)> <zone size (MiB)> :\n"
	       "      Specify the percentage of the capacity to use for\n"
	       "      conventional zones and the size in MiB of zones\n",
	       basename(prog));

	return 1;
}

int main(int argc, char **argv)
{
	struct zbc_device_info info;
	struct zbc_device *dev;
	long long conv_num, conv_sz, zone_sz;
	double conv_p;
	int i, ret = -1;
	char *path;

	/* Check command line */
	if (argc < 5)
		return zbc_set_zones_usage(argv[0]);

	/* Parse options */
	for (i = 1; i < argc; i++) {

		if (strcmp(argv[i], "-h") == 0 ||
		    strcmp(argv[i], "--help") == 0)
			return zbc_set_zones_usage(argv[0]);

		if (strcmp(argv[i], "-v") == 0) {

			zbc_set_log_level("debug");

		} else if (argv[i][0] == '-') {

			fprintf(stderr, "Unknown option \"%s\"\n", argv[i]);
			return 1;

		} else {

			break;

		}

	}

	if (i > argc - 3) {
		fprintf(stderr, "Invalid command line\n");
		return 1;
	}

	/* Open device: only allow fake device backend driver */
	path = argv[i];
	ret = zbc_open(path, O_RDWR | ZBC_O_DRV_FAKE | ZBC_O_SETZONES, &dev);
	if (ret < 0) {
		if (ret == -ENODEV)
			fprintf(stderr, "Unsupported device type\n");
		else
			fprintf(stderr, "Open %s failed (%s)\n",
				path, strerror(-ret));
		return 1;
	}

	/* Get device info */
	zbc_get_device_info(dev, &info);
	if (info.zbd_type != ZBC_DT_FAKE) {
		fprintf(stderr,
			"The fake backend driver is not in use for device %s\n",
			path);
		ret = 1;
		goto out;
	}

	printf("Device %s:\n", path);
	zbc_print_device_info(&info, stdout);
	printf("\n");

	/* Process command */
	printf("Setting zones:\n");
	i++;

	if (strcmp(argv[i], "set_sz") == 0) {

		/*
		 * Set conventional zones capacity and zone size for all zones.
		 */
		if (i != argc - 3) {
			fprintf(stderr,
				"Invalid number of arguments for set_sz\n");
			goto out;
		}

		/* Get arguments */
		conv_sz = (strtoll(argv[i + 1], NULL, 10) * 1024 * 1024) >> 9;
		if (conv_sz < 0) {
			fprintf(stderr, "Invalid conventional zones size %s\n",
				argv[i + 1]);
			ret = 1;
			goto out;
		}

		zone_sz = (strtoll(argv[i + 2], NULL, 10) * 1024 * 1024)
			>> 9;
		if (zone_sz <= 0) {
			fprintf(stderr, "Invalid zone size %s\n",
				argv[i + 2]);
			ret = 1;
			goto out;
		}

	} else if (strcmp(argv[i], "set_ps") == 0) {

		/*
		 * Set conventional zones capacity percentage and
		 * zone size for all zones.
		 */
		if (i != argc - 3) {
			fprintf(stderr,
				"Invalid number of arguments for set_ps\n");
			goto out;
		}

		/* Get arguments */
		conv_p = strtof(argv[i + 1], NULL);
		if (conv_p < 0 || conv_p >= 100.0) {
			fprintf(stderr, "Invalid capacity percentage %s for conventional zones\n",
				argv[i + 1]);
			ret = 1;
			goto out;
		}

		zone_sz = (strtoll(argv[i + 2], NULL, 10) * 1024ULL * 1024ULL)
			>> 9;
		if (zone_sz <= 0) {
			fprintf(stderr, "Invalid zone size %s\n",
				argv[i + 2]);
			ret = 1;
			goto out;
		}

		conv_sz = (long long)((double) info.zbd_sectors * (double) conv_p) / 100;
		if (conv_p && conv_sz < zone_sz)
			conv_sz = zone_sz;

	} else {

		fprintf(stderr, "Unknown command \"%s\"\n", argv[i]);
		goto out;

	}

	if ((unsigned long long)conv_sz >= info.zbd_sectors) {
		fprintf(stderr,
			"Invalid conventional zone capacity (too large)\n");
		ret = 1;
		goto out;
	}
	if (conv_sz && conv_sz < zone_sz) {
		fprintf(stderr,
			"Invalid conventional zone capacity (too low)\n");
		ret = 1;
		goto out;
	}

	if (conv_sz) {
		conv_num = conv_sz / zone_sz;
		if (!conv_num || conv_sz % zone_sz)
			conv_num++;
		conv_sz = zone_sz * conv_num;
	} else {
		conv_num = 0;
	}

	printf("    Zone size: %lld MiB (%lld sectors)\n",
	       (zone_sz << 9) / (1024 * 1024),
	       zone_sz);

	printf("    Conventional zones: %lld MiB (%lld sectors), %.02F %% of total capacity), %lld zones\n",
	       (conv_sz << 9) / (1024 * 1024),
	       conv_sz,
	       100.0 * (double)conv_sz / (double)info.zbd_sectors,
	       conv_num);

	printf("    Sequential zones: %llu zones\n",
	       (info.zbd_sectors - conv_sz) / zone_sz);

	ret = zbc_set_zones(dev, conv_sz, zone_sz);
	if (ret != 0) {
		fprintf(stderr,
			"zbc_set_zones failed %d (%s)\n",
			ret,
			strerror(-ret));
		ret = 1;
	}

out:
	zbc_close(dev);

	return ret;
}

